
#Progetto per l'esame di Foundation of computer science


## Progetto finale <br/> Marco Martines - Matricola 873849 <br/> Geralda Nushi - Matricola 873269 <br/> Jacopo Fantini - Matricola 864115






Importiamo le librerie RE e Pandas. 
Re viene utilizzato nella normalizzazione per trovare i match tra il file school and country e i 3 dataset in modo da normalizzare tutti i nomi delle università

In [ ]:
import pandas as pd
import re
import numpy as np

#Importiamo i dataset e eseguiamo un preprocessing

In [ ]:
shanghai_data="https://raw.githubusercontent.com/martibic/focs/main/shanghaiData.csv"
shanghai_ranking = pd.read_csv(shanghai_data)
shanghai_ranking.head(3)

world_rank          university_name national_rank  ...    pub   pcp  year
0          1       Harvard University             1  ...  100.0  72.4  2005
1          2  University of Cambridge             1  ...   70.9  66.9  2005
2          3      Stanford University             2  ...   72.3  65.0  2005

[3 rows x 11 columns]

In questo passaggio analizziamo il dataset per tipologia dei campi e la presenza di eventuali valori nulli, ci interessa cancellare i valori nulli presenti nella colonna UNIVERSITY NAME

In [ ]:
print(str(shanghai_ranking.dtypes)+"\n")
print(str(shanghai_ranking.isnull().any())+"\n")
print("Number of rows = " + str(len(shanghai_ranking)))
print("Number University name NULL = " + str(len(shanghai_ranking[shanghai_ranking['university_name'].isnull()==True])))
print("Number University unique = " + str(len(shanghai_ranking['university_name'].unique())))

world_rank          object
university_name     object
national_rank       object
total_score        float64
alumni             float64
award              float64
hici               float64
ns                 float64
pub                float64
pcp                float64
year                 int64
dtype: object

world_rank         False
university_name     True
national_rank       True
total_score         True
alumni              True
award               True
hici                True
ns                  True
pub                 True
pcp                 True
year               False
dtype: bool

Number of rows = 4897
Number University name NULL = 1
Number University unique = 659


Eliminiamo le righe nulle del campo UNIVERSITY NAME, resettiamo l'indice virtuale e mostriamo la tabella.
Le righe contenute nel dataset shanghai_ranking erano 4897, quindi cancello solo la riga dove university_name è NULL

In [ ]:
shanghai_ranking.dropna(subset=['university_name'],inplace=True)
shanghai_ranking.iloc(shanghai_ranking.reset_index(drop=True,inplace=True))
shanghai_ranking

world_rank                              university_name  ...   pcp  year
0             1                           Harvard University  ...  72.4  2005
1             2                      University of Cambridge  ...  66.9  2005
2             3                          Stanford University  ...  65.0  2005
3             4           University of California, Berkeley  ...  52.7  2005
4             5  Massachusetts Institute of Technology (MIT)  ...  53.0  2005
...         ...                                          ...  ...   ...   ...
4891    401-500                        University of Trieste  ...  20.1  2015
4892    401-500                       University of Zaragoza  ...  13.1  2015
4893    401-500                        Utah State University  ...  15.5  2015
4894    401-500              Vienna University of Technology  ...  22.9  2015
4895    401-500                       Wake Forest University  ...  11.9  2015

[4896 rows x 11 columns]

In [ ]:
times_data="https://raw.githubusercontent.com/martibic/focs/main/timesData.csv"
times_ranking = pd.read_csv(times_data)
times_ranking.head(10)

world_rank                        university_name  ... female_male_ratio  year
0          1                     Harvard University  ...               NaN  2011
1          2     California Institute of Technology  ...           33 : 67  2011
2          3  Massachusetts Institute of Technology  ...           37 : 63  2011
3          4                    Stanford University  ...           42 : 58  2011
4          5                   Princeton University  ...           45 : 55  2011
5          6                University of Cambridge  ...           46 : 54  2011
6          6                   University of Oxford  ...           46 : 54  2011
7          8     University of California, Berkeley  ...           50 : 50  2011
8          9                Imperial College London  ...           37 : 63  2011
9         10                        Yale University  ...           50 : 50  2011

[10 rows x 14 columns]

In [ ]:
print(str(times_ranking.dtypes)+"\n")
print(str(times_ranking.isnull().any())+"\n")
print("Number of rows = " + str(len(times_ranking)))
print("Number University name NULL = " + str(len(times_ranking[times_ranking['university_name'].isnull()==True])))
print("Number University unique = " + str(len(times_ranking['university_name'].unique())))

world_rank                 object
university_name            object
country                    object
teaching                  float64
international              object
research                  float64
citations                 float64
income                     object
total_score                object
num_students               object
student_staff_ratio       float64
international_students     object
female_male_ratio          object
year                        int64
dtype: object

world_rank                False
university_name           False
country                   False
teaching                  False
international             False
research                  False
citations                 False
income                    False
total_score               False
num_students               True
student_staff_ratio        True
international_students     True
female_male_ratio          True
year                      False
dtype: bool

Number of rows = 2603
Number University name NUL

In [ ]:
world_data="https://raw.githubusercontent.com/martibic/focs/main/cwurData.csv"
world_ranking = pd.read_csv(world_data)
world_ranking.head(3)

world_rank                            institution  ...   score  year
0           1                     Harvard University  ...  100.00  2012
1           2  Massachusetts Institute of Technology  ...   91.67  2012
2           3                    Stanford University  ...   89.50  2012

[3 rows x 14 columns]

Rinominiamo la colonna INSTITUTION nel dataset world_ranking in modo da standardizzare come negli altri dataset

In [ ]:
world_ranking.rename(columns={'institution':'university_name'},inplace=True)
world_ranking.head(3)

world_rank                        university_name  ...   score  year
0           1                     Harvard University  ...  100.00  2012
1           2  Massachusetts Institute of Technology  ...   91.67  2012
2           3                    Stanford University  ...   89.50  2012

[3 rows x 14 columns]

In [ ]:
print(str(world_ranking.dtypes)+"\n")
print(str(world_ranking.isnull().any())+"\n")
print("Number of rows = " + str(len(world_ranking)))
print("Number University name NULL = " + str(len(world_ranking[world_ranking['university_name'].isnull()==True])))
print("Number University unique = " + str(len(world_ranking['university_name'].unique())))

world_rank                int64
university_name          object
country                  object
national_rank             int64
quality_of_education      int64
alumni_employment         int64
quality_of_faculty        int64
publications              int64
influence                 int64
citations                 int64
broad_impact            float64
patents                   int64
score                   float64
year                      int64
dtype: object

world_rank              False
university_name         False
country                 False
national_rank           False
quality_of_education    False
alumni_employment       False
quality_of_faculty      False
publications            False
influence               False
citations               False
broad_impact             True
patents                 False
score                   False
year                    False
dtype: bool

Number of rows = 2200
Number University name NULL = 0
Number University unique = 1024


In [ ]:
school_and_country = pd.read_csv("https://raw.githubusercontent.com/martibic/focs/main/school_and_country_table.csv"
)
school_and_country.rename(columns={'school_name':'uni_name_norm'},inplace=True)
school_and_country.head(3)

uni_name_norm                   country
0                     Harvard University  United States of America
1     California Institute of Technology  United States of America
2  Massachusetts Institute of Technology  United States of America

In [ ]:
print(str(school_and_country.dtypes)+"\n")
print(str(school_and_country.isnull().any())+"\n")
print("Number of rows = " + str(len(school_and_country)))
print("Number University name NULL = " + str(len(school_and_country[school_and_country['uni_name_norm'].isnull()==True])))
print("Number University unique = " + str(len(school_and_country['uni_name_norm'].unique())))

uni_name_norm    object
country          object
dtype: object

uni_name_norm    False
country          False
dtype: bool

Number of rows = 818
Number University name NULL = 0
Number University unique = 818


Creazione di una funzione di normalizzazione dei nomi delle università prendendo come punto di riferimento i nome delle università contenute in School and country, nel primo **if** prendiamo i valori uguali tra i 2 dataset, in caso di non compatibilità si verificano 2 casi. Nel primo cerchiamo i nome delle università del dataset da normalizzare contenuti nel dataset school and country e nel secondo l'opposto.

In [ ]:
def name_normalized (column_name):
    
    for university in school_and_country['uni_name_norm']:
      if column_name == university:
        return university
      else:
        pattern = re.compile(column_name)
        occurrence = pattern.search(university)
        if occurrence: 
          return university
        else: 
          pattern2 = re.compile(university)
          occurence2 = pattern2.search(column_name)
          if occurence2:
            return university

In [ ]:
shanghai_ranking['uni_name_norm'] = shanghai_ranking['university_name'].apply(name_normalized)

In [ ]:
shanghai_ranking[shanghai_ranking['uni_name_norm'].isnull()==True]

world_rank  ... uni_name_norm
17           18  ...          None
22           23  ...          None
27           28  ...          None
29           30  ...          None
34           35  ...          None
...         ...  ...           ...
4883    401-500  ...          None
4884    401-500  ...          None
4885    401-500  ...          None
4886    401-500  ...          None
4893    401-500  ...          None

[1207 rows x 12 columns]

Sistemiamo i valori None con i valori originali

In [ ]:
def none_value (db):
  norm = db['uni_name_norm']
  norm1 = db['university_name']
  if norm==None:
    return norm1
  else:
    return norm

In [ ]:
shanghai_ranking["uni_name_norm"]=shanghai_ranking.apply(none_value, axis=1)

In [ ]:
shanghai_ranking[shanghai_ranking['uni_name_norm'].isnull()==True]

Empty DataFrame
Columns: [world_rank, university_name, national_rank, total_score, alumni, award, hici, ns, pub, pcp, year, uni_name_norm]
Index: []

In [ ]:
shanghai_ranking.head(10)

world_rank  ...                          uni_name_norm
0          1  ...                     Harvard University
1          2  ...                University of Cambridge
2          3  ...                    Stanford University
3          4  ...     University of California, Berkeley
4          5  ...  Massachusetts Institute of Technology
5          6  ...     California Institute of Technology
6          7  ...                    Columbia University
7          8  ...                   Princeton University
8          9  ...                  University of Chicago
9         10  ...                   University of Oxford

[10 rows x 12 columns]

In [ ]:
world_ranking['uni_name_norm'] = world_ranking['university_name'].apply(name_normalized)

In [ ]:
world_ranking["uni_name_norm"]=world_ranking.apply(none_value, axis=1)

In [ ]:
world_ranking[world_ranking['uni_name_norm'].isnull()==True]

Empty DataFrame
Columns: [world_rank, university_name, country, national_rank, quality_of_education, alumni_employment, quality_of_faculty, publications, influence, citations, broad_impact, patents, score, year, uni_name_norm]
Index: []

In [ ]:
world_ranking.head(10)

world_rank  ...                          uni_name_norm
0           1  ...                     Harvard University
1           2  ...  Massachusetts Institute of Technology
2           3  ...                    Stanford University
3           4  ...                University of Cambridge
4           5  ...     California Institute of Technology
5           6  ...                   Princeton University
6           7  ...                   University of Oxford
7           8  ...                        Yale University
8           9  ...                    Columbia University
9          10  ...     University of California, Berkeley

[10 rows x 15 columns]

In [ ]:
times_ranking['uni_name_norm'] = times_ranking['university_name'].apply(name_normalized)

In [ ]:
times_ranking["uni_name_norm"]=times_ranking.apply(none_value, axis=1)

In [ ]:
times_ranking[times_ranking['uni_name_norm'].isnull()==True]

Empty DataFrame
Columns: [world_rank, university_name, country, teaching, international, research, citations, income, total_score, num_students, student_staff_ratio, international_students, female_male_ratio, year, uni_name_norm]
Index: []

In [ ]:
times_ranking.head(10)

world_rank  ...                          uni_name_norm
0          1  ...                     Harvard University
1          2  ...     California Institute of Technology
2          3  ...  Massachusetts Institute of Technology
3          4  ...                    Stanford University
4          5  ...                   Princeton University
5          6  ...                University of Cambridge
6          6  ...                   University of Oxford
7          8  ...     University of California, Berkeley
8          9  ...                Imperial College London
9         10  ...                        Yale University

[10 rows x 15 columns]

#1. For each university, extract from the times dataset the most recent and the least recent data, obtaining two separate dataframes

In [ ]:
t_max = times_ranking.iloc[times_ranking.groupby('uni_name_norm')['year'].idxmax()].copy()
t_max.head()

world_rank  ...                             uni_name_norm
2405    601-800  ...  AGH University of Science and Technology
2003    201-250  ...                        Aalborg University
2056    251-300  ...                          Aalto University
1908       =106  ...                         Aarhus University
2105    301-350  ...                    Aberystwyth University

[5 rows x 15 columns]

In [ ]:
t_min = times_ranking.iloc[times_ranking.groupby('uni_name_norm')['year'].idxmin()].copy()
t_min.head()

world_rank  ...                             uni_name_norm
2405    601-800  ...  AGH University of Science and Technology
501     301-350  ...                        Aalborg University
502     301-350  ...                          Aalto University
166         167  ...                         Aarhus University
476     276-300  ...                    Aberystwyth University

[5 rows x 15 columns]

#2. For each university, compute the improvement in income between the least recent and the most recent data points

In [ ]:
t_min_max = pd.merge(t_min, t_max, on = ['uni_name_norm'],how='inner',suffixes=('_min', '_max'))\
[['uni_name_norm','income_min','income_max','year_min','year_max']]
t_min_max

uni_name_norm income_min  ... year_min  year_max
0    AGH University of Science and Technology          -  ...     2016      2016
1                          Aalborg University       36.4  ...     2012      2016
2                            Aalto University       61.9  ...     2012      2016
3                           Aarhus University       61.5  ...     2011      2016
4                      Aberystwyth University       35.5  ...     2012      2016
..                                        ...        ...  ...      ...       ...
791               Yıldız Technical University       44.0  ...     2016      2016
792                       Zhejiang University       70.3  ...     2011      2016
793                  École Normale Supérieure       30.7  ...     2011      2016
794                       École Polytechnique          -  ...     2011      2016
795                         Örebro University       29.8  ...     2016      2016

[796 rows x 5 columns]

Analizziamo i 3 dataset per vedere se è stato eseguito il join tra tutti i valori

In [ ]:
print(str(t_min_max.dtypes)+"\n")
print("Number of rows max dataset " + str(len(t_max)))
print("Number of rows min dataset " + str(len(t_min)))
print("Number of rows join dataset " + str(len(t_min_max)))

uni_name_norm    object
income_min       object
income_max       object
year_min          int64
year_max          int64
dtype: object

Number of rows max dataset 796
Number of rows min dataset 796
Number of rows join dataset 796


income_max e income_min sono vengono riconosciuti come float64 a causa del carattere - trasformiamo tutto in numero tramite to_numeric trasformando le eccezioni in NULL con error = ‘coerce’

In [ ]:
t_min_max['income_min'] = pd.to_numeric(t_min_max['income_min'], errors = 'coerce')
t_min_max['income_max'] = pd.to_numeric(t_min_max['income_max'], errors = 'coerce')

t_min_max.dtypes

uni_name_norm     object
income_min       float64
income_max       float64
year_min           int64
year_max           int64
dtype: object

In [ ]:
t_min_max['delta'] = t_min_max['income_max'] - t_min_max['income_min']
t_min_max

uni_name_norm  income_min  ...  year_max  delta
0    AGH University of Science and Technology         NaN  ...      2016    NaN
1                          Aalborg University        36.4  ...      2016    7.3
2                            Aalto University        61.9  ...      2016   -0.3
3                           Aarhus University        61.5  ...      2016    6.8
4                      Aberystwyth University        35.5  ...      2016   -4.2
..                                        ...         ...  ...       ...    ...
791               Yıldız Technical University        44.0  ...      2016    0.0
792                       Zhejiang University        70.3  ...      2016   25.9
793                  École Normale Supérieure        30.7  ...      2016    6.4
794                       École Polytechnique         NaN  ...      2016    NaN
795                         Örebro University        29.8  ...      2016    0.0

[796 rows x 6 columns]

# 3. Find the university with the largest increase computed in the previous point

confrontiamo il massimo con tutti i valori della colonna delta per printare l'università con il massimo delta

In [ ]:
t_min_max[t_min_max['delta'] == t_min_max['delta'].max()] 

uni_name_norm  income_min  income_max  year_min  year_max  delta
417    TU Dresden        31.9        99.7      2012      2016   67.8

#4. For each ranking, consider only the most recent data point. For each university, compute the maximum difference between the rankings (e.g. for Aarhus University the value is 122-73=49). Notice that some rankings are expressed as a range

##Raggruppiamo per università estraendo quelle con il numero di anno maggiore per i rimanenti 2 dataset, time raking è stato fatto nel punto 3

In [ ]:
w_max = world_ranking.iloc[world_ranking.groupby('uni_name_norm')['year'].idxmax()].copy()
w_max.head(3)

world_rank  ...                             uni_name_norm
1981         782  ...  AGH University of Science and Technology
1764         565  ...                        Aalborg University
1620         421  ...                          Aalto University

[3 rows x 15 columns]

In [ ]:
s_max = shanghai_ranking.iloc[shanghai_ranking.groupby('uni_name_norm')['year'].idxmax()].copy()
s_max.head(3)

world_rank     university_name  ...  year       uni_name_norm
4696    301-400  Aalborg University  ...  2015  Aalborg University
4796    401-500    Aalto University  ...  2015    Aalto University
4468         73   Aarhus University  ...  2015   Aarhus University

[3 rows x 12 columns]

##Uniamo i 3 dataset con solo l'anno massimo tramite la chiave che abbiamo creato con la normalizzazione

In [ ]:
ranking_temp = pd.merge(t_max[['world_rank','uni_name_norm']],s_max[['world_rank','uni_name_norm']], on ='uni_name_norm')
ranking = pd.merge(ranking_temp[['world_rank_x','world_rank_y', 'uni_name_norm']],w_max[['world_rank','uni_name_norm']], on ='uni_name_norm')
ranking.head(3)

world_rank_x world_rank_y       uni_name_norm  world_rank
0      201-250      301-400  Aalborg University         565
1      251-300      401-500    Aalto University         421
2         =106           73   Aarhus University         122

##Rinominiamo le colonne per capire a quale dataset appartiengono

In [ ]:
ranking.rename(columns={'world_rank_x':'times_ranking',\
                                                'world_rank_y':'shanghai_ranking',
                                                'world_rank':'world_ranking'},inplace=True)
ranking = ranking[['uni_name_norm','times_ranking','shanghai_ranking','world_ranking']]
ranking.head(3)

uni_name_norm times_ranking shanghai_ranking  world_ranking
0  Aalborg University       201-250          301-400            565
1    Aalto University       251-300          401-500            421
2   Aarhus University          =106               73            122

## Definiamo la funzione, che ritorna il valore medio della classe nel caso in sia un range altrimenti toglie l'uguale laddove sia presente oppure ritorna il valore senza le precedenti condizioni. Tutti i valori vengono convertiti in INT

In [ ]:
def rank_normalized(value):
  if '-' in value:
    pattern = re.compile("(\d+)\-(\d+)")
    pat = pattern.search(value)
    rank = (int(pat.group(1))+int(pat.group(2)))/2
    return int(rank) 
  elif '=' in value:
    rank2=value.replace('=','')
    return int(rank2) 
  else:
    return int(value)

In [ ]:
ranking['times_ranking'] = ranking['times_ranking'].apply(rank_normalized)
ranking['shanghai_ranking'] = ranking['shanghai_ranking'].apply(rank_normalized)
ranking.head(3)

uni_name_norm  times_ranking  shanghai_ranking  world_ranking
0  Aalborg University            225               350            565
1    Aalto University            275               450            421
2   Aarhus University            106                73            122

In [ ]:
ranking.isnull().any()

uni_name_norm       False
times_ranking       False
shanghai_ranking    False
world_ranking       False
dtype: bool

##Definiamo la funzione che inizialmente trova il valore massimo e minimo per ogni università, per calcolare la differenza massima.

In [ ]:
def diff_max(db):
  massimo = max(db['times_ranking'],db['shanghai_ranking'],db['world_ranking'])
  minimo = min(db['times_ranking'],db['shanghai_ranking'],db['world_ranking'])
  return (massimo-minimo)

In [ ]:
ranking['max_difference'] = ranking.apply(diff_max, axis=1) #L'axis 1 agirà su tutte le COLONNE di ogni RIGA
ranking.head(3)

uni_name_norm  times_ranking  ...  world_ranking  max_difference
0  Aalborg University            225  ...            565             340
1    Aalto University            275  ...            421             175
2   Aarhus University            106  ...            122              49

[3 rows x 5 columns]

# 5. Consider only the most recent data point of the times dataset. Compute the number of male and female students for each country

## 1) splittiamo i valori che contengono :
## 2) assegnamo al primo valore dello split il gruppo femmine e al secondo i maschi, trasformiamo in numero e cancelliamo le eccezioni
## 3) rimuoviamo la virgola dal campo '	num_students ' per renderla un float
## 4) Calcoliamo il numero di maschi e femmine in base al totale
## 5) Raggruppiamo i risultati per country con aggregazione di tipo somma

In [ ]:
split_fm = t_max['female_male_ratio'].str.split(':', 1, expand=True) 
df_split = t_max.assign(female=split_fm[0], male=split_fm[1]) 
df_split['female'] = pd.to_numeric(df_split['female'], errors = 'coerce').astype(float)
df_split['male'] = pd.to_numeric(df_split['male'], errors = 'coerce').astype(float)
df_split['num_students'] = df_split['num_students'].str.replace(',','').astype(float)
df_split['female_num'] = df_split['num_students']*df_split['female']/100
df_split['male_num'] = df_split['num_students']*df_split['male']/100
gender_num = df_split[df_split['male_num'].isnull()==False].copy()
final = gender_num.groupby('country',as_index=False)[['female_num','male_num']].sum()
final

country  female_num    male_num
0                  Argentina    67191.26    41181.74
1                  Australia   391735.88   321640.12
2                    Austria    64651.08    62399.92
3                 Bangladesh    21323.44    41392.56
4                    Belarus    20219.07     9083.93
..                       ...         ...         ...
64                    Uganda    18670.00    18670.00
65                   Ukraine    17846.00    19250.00
66      United Arab Emirates     7007.96     2213.04
67            United Kingdom   711813.62   613028.38
68  United States of America  1589731.79  1545501.21

[69 rows x 3 columns]

# 6. Find the universities where the ratio between female and male is below the average ratio (computed over all universities)

In [ ]:
tot_f = gender_num['female_num'].sum()
tot_m =gender_num['male_num'].sum()
mean = tot_f/tot_m
gender_num['diff']=round(gender_num['female_num'] / gender_num['male_num'],1)
t_b_ave = gender_num.loc[gender_num['diff'] < mean].copy()
t_b_ave

world_rank                           university_name  ...  male_num  diff
2003    201-250                        Aalborg University  ...   9059.44   0.9
2056    251-300                          Aalto University  ...  10947.32   0.5
2105    301-350                    Aberystwyth University  ...   4811.04   0.9
2406    601-800                           Ajou University  ...   8513.02   0.5
2408    601-800                     Alexandria University  ...  68812.74   0.9
...         ...                                       ...  ...       ...   ...
2602    601-800                        Yuan Ze University  ...   4937.91   0.8
2599    601-800               Yıldız Technical University  ...  20011.52   0.6
2104    251-300                       Zhejiang University  ...  28029.72   0.7
1856         54                  École Normale Supérieure  ...   1296.00   0.9
1833         31  École Polytechnique Fédérale de Lausanne  ...   7056.18   0.4

[334 rows x 20 columns]

# 7. For each country, compute the fraction of the students in the country that are in one of the universities computed in the previous point (that is, the denominator of the ratio is the total number of students over all universities in the country)

Partiamo raggrupando le nazioni presenti nella tabella t_b_ave, per ciascuna calcoliamo la somma del numero di studenti frequentanti le università con sede in quel determinato paese considerato.

In [ ]:
students_country_b_ave = t_b_ave.groupby('country',as_index=False)['num_students'].sum()
students_country_b_ave

country  num_students
0                  Australia      146235.0
1                    Austria       53607.0
2                 Bangladesh       62716.0
3                     Brazil      278774.0
4                     Canada       86779.0
5                      Chile       61744.0
6                      China      835033.0
7                   Colombia       54645.0
8             Czech Republic       38685.0
9                    Denmark       45705.0
10                     Egypt      359372.0
11                   Estonia       14061.0
12                   Finland       39788.0
13                    France      151506.0
14                   Germany      435244.0
15                     Ghana       37274.0
16                    Greece       57198.0
17                 Hong Kong       22064.0
18                   Hungary       22447.0
19                     India      155370.0
20                 Indonesia       43565.0
21                      Iran      107280.0
22                    Israel       30550.0
23                     Italy       78762.0
24                     Japan      447415.0
25                     Kenya       72676.0
26                Luxembourg        5144.0
27                  Malaysia       16270.0
28                    Mexico      151731.0
29                   Morocco       56060.0
30               Netherlands       58883.0
31               New Zealand       12187.0
32                   Nigeria       26247.0
33                    Norway       17381.0
34                      Oman        7089.0
35                  Pakistan        9808.0
36                    Poland       59789.0
37                  Portugal       41436.0
38       Republic of Ireland       34597.0
39        Russian Federation      119356.0
40                 Singapore       56620.0
41                  Slovakia       16400.0
42               South Korea      276058.0
43                     Spain      102916.0
44                    Sweden       20667.0
45               Switzerland       34692.0
46                    Taiwan      202503.0
47                  Thailand       66058.0
48                    Turkey      216017.0
49                    Uganda       37340.0
50                   Ukraine       22686.0
51            United Kingdom      259977.0
52  United States of America     1726340.0

Per ogni paese Country della tabella 'students_country' calcoliamo il numero totale di studenti , maschi e femmine.

In [ ]:
final=final.reset_index()

final['tot_stud']= final['female_num']+final['male_num']
students_country = final[['country','tot_stud']]
students_country

country   tot_stud
0                  Argentina   108373.0
1                  Australia   713376.0
2                    Austria   127051.0
3                 Bangladesh    62716.0
4                    Belarus    29303.0
..                       ...        ...
64                    Uganda    37340.0
65                   Ukraine    37096.0
66      United Arab Emirates     9221.0
67            United Kingdom  1324842.0
68  United States of America  3135233.0

[69 rows x 2 columns]

Uniamo le tabelle 'students_country_b_ave' e 'students_country' sulla colonna 'country'. Calcoliamo poi la percentuale tra 'num_students' e 'tot_stud'

In [ ]:
students_country_ratio=pd.merge(students_country_b_ave,students_country, on='country')
students_country_ratio['ratio_b_ave']=round(students_country_ratio['num_students']/students_country_ratio['tot_stud']*100)
students_country_ratio

country  num_students   tot_stud  ratio_b_ave
0                  Australia      146235.0   713376.0         20.0
1                    Austria       53607.0   127051.0         42.0
2                 Bangladesh       62716.0    62716.0        100.0
3                     Brazil      278774.0   494251.0         56.0
4                     Canada       86779.0   555134.0         16.0
5                      Chile       61744.0    61744.0        100.0
6                      China      835033.0   964890.0         87.0
7                   Colombia       54645.0    54645.0        100.0
8             Czech Republic       38685.0   175871.0         22.0
9                    Denmark       45705.0   119355.0         38.0
10                     Egypt      359372.0   376189.0         96.0
11                   Estonia       14061.0    24330.0         58.0
12                   Finland       39788.0   106165.0         37.0
13                    France      151506.0   645841.0         23.0
14                   Germany      435244.0   918502.0         47.0
15                     Ghana       37274.0    37274.0        100.0
16                    Greece       57198.0   202013.0         28.0
17                 Hong Kong       22064.0    52340.0         42.0
18                   Hungary       22447.0    87765.0         26.0
19                     India      155370.0   155370.0        100.0
20                 Indonesia       43565.0    43565.0        100.0
21                      Iran      107280.0   138992.0         77.0
22                    Israel       30550.0    83423.0         37.0
23                     Italy       78762.0  1009470.0          8.0
24                     Japan      447415.0   467844.0         96.0
25                     Kenya       72676.0    72676.0        100.0
26                Luxembourg        5144.0     5144.0        100.0
27                  Malaysia       16270.0   161827.0         10.0
28                    Mexico      151731.0   151731.0        100.0
29                   Morocco       56060.0    56060.0        100.0
30               Netherlands       58883.0   225877.0         26.0
31               New Zealand       12187.0   125957.0         10.0
32                   Nigeria       26247.0    26247.0        100.0
33                    Norway       17381.0    39402.0         44.0
34                      Oman        7089.0     7089.0        100.0
35                  Pakistan        9808.0    22496.0         44.0
36                    Poland       59789.0   225313.0         27.0
37                  Portugal       41436.0   116041.0         36.0
38       Republic of Ireland       34597.0    90116.0         38.0
39        Russian Federation      119356.0   208613.0         57.0
40                 Singapore       56620.0    56620.0        100.0
41                  Slovakia       16400.0    37279.0         44.0
42               South Korea      276058.0   293683.0         94.0
43                     Spain      102916.0   749668.0         14.0
44                    Sweden       20667.0   132622.0         16.0
45               Switzerland       34692.0   118976.0         29.0
46                    Taiwan      202503.0   289505.0         70.0
47                  Thailand       66058.0   210122.0         31.0
48                    Turkey      216017.0   639342.0         34.0
49                    Uganda       37340.0    37340.0        100.0
50                   Ukraine       22686.0    37096.0         61.0
51            United Kingdom      259977.0  1324842.0         20.0
52  United States of America     1726340.0  3135233.0         55.0

# 8. Read the file educational_attainment_supplementary_data.csv, discarding any row with missing country_name or series_name

##Importiamo la tabella educational_attainment_supplementary_data.csv 

In [ ]:
educational_attainment=pd.read_csv("https://raw.githubusercontent.com/martibic/focs/main/educational_attainment_supplementary_data.csv")
educational_attainment

country_name  ... 2015
0                                            Afghanistan  ...  NaN
1                                            Afghanistan  ...  NaN
2                                            Afghanistan  ...  NaN
3                                            Afghanistan  ...  NaN
4                                            Afghanistan  ...  NaN
...                                                  ...  ...  ...
79050                                                NaN  ...  NaN
79051                                                NaN  ...  NaN
79052                                                NaN  ...  NaN
79053  Data from database: Education Statistics: Educ...  ...  NaN
79054                           Last Updated: 10/20/2015  ...  NaN

[79055 rows x 29 columns]

##Controlliamo per tutte le righe della tabella se vi sono presenti valori null in corrispondenza delle colonne 'country_name' e 'series_name'

In [ ]:
educational_attainment[(educational_attainment['country_name'].isnull()==True)|(educational_attainment['series_name'].isnull()==True)]

country_name  ... 2015
79050                                                NaN  ...  NaN
79051                                                NaN  ...  NaN
79052                                                NaN  ...  NaN
79053  Data from database: Education Statistics: Educ...  ...  NaN
79054                           Last Updated: 10/20/2015  ...  NaN

[5 rows x 29 columns]

##Tramite la funzione dropna eliminiamo i valori nulli nelle specifiche colonne selezionate e visualizziamo il dataset senza valori nulli come da richiesta

In [ ]:
educational_attainment.dropna(subset=['country_name'],inplace=True)
educational_attainment.dropna(subset=['series_name'],inplace=True)
educational_attainment.iloc(educational_attainment.reset_index(drop=True,inplace=True))
educational_attainment

country_name  ... 2015
0      Afghanistan  ...  NaN
1      Afghanistan  ...  NaN
2      Afghanistan  ...  NaN
3      Afghanistan  ...  NaN
4      Afghanistan  ...  NaN
...            ...  ...  ...
79045     Zimbabwe  ...  NaN
79046     Zimbabwe  ...  NaN
79047     Zimbabwe  ...  NaN
79048     Zimbabwe  ...  NaN
79049     Zimbabwe  ...  NaN

[79050 rows x 29 columns]

# 9. From attainment build a dataframe with the same data, but with 4 columns: country_name, series_name, year, value


Scomponiamo il dataframe 'educational_attainment' aggregato, in dati non aggregati. Creiamo una tabella dal nome 'pivot', che per  ogni riga stampa il valore di 'country_name' e di 'series_name', l'anno di riferimento della singola osservazione e il valore che la colonna 'series_name' assume in quell'anno.

In [ ]:
pivot = pd.melt(educational_attainment, \
                             id_vars =['country_name','series_name'], value_vars =['1985','1986','1987','1990','1991','1992','1993','1995','1996','1997','1998','1999','2000','2001','2002','2003','2004','2005','2006','2007','2008','2009','2010','2011','2012','2013','2015'], \
                             var_name ='years', \
                             value_name ='values')
pivot[(pivot['values'].isnull()==False)]

country_name  ...   values
0               Afghanistan  ...  0.33000
1               Afghanistan  ...  1.03000
2               Afghanistan  ...  0.83000
3               Afghanistan  ...  2.34000
4               Afghanistan  ...  0.54000
...                     ...  ...      ...
2054020  West Bank and Gaza  ...  1.48356
2054021  West Bank and Gaza  ...  1.88820
2054022  West Bank and Gaza  ...  0.00000
2054023  West Bank and Gaza  ...  0.00000
2054024  West Bank and Gaza  ...  0.00000

[318290 rows x 4 columns]

# 10. For each university, find the number of rankings in which they appear (it suffices to appear in one year for each ranking)¶


Deteriniamo quante volte una data università presente nei tre datasets times_ranking, shanghai_ranking, world_ranking si ripete all'interno dello stesso dataset, determinando così il numero di classificazioni richiesto per ciascuna università.

In [ ]:
uni_appearance = {}
for university in school_and_country['uni_name_norm']:
  count = 0
  for uni in times_ranking['university_name']:
    if uni == university:
      count += 1
      break
  for uni in shanghai_ranking['university_name']:
    if uni == university:
      count += 1
      break
  for uni in world_ranking['university_name']:
    if uni == university:
      count += 1
      break
  uni_appearance.setdefault(university,count)
  

uni_appearance

{'AGH University of Science and Technology': 2,
 'Aalborg University': 3,
 'Aalto University': 3,
 'Aarhus University': 3,
 'Aberystwyth University': 2,
 'Adam Mickiewicz University': 1,
 'Aix-Marseille University': 3,
 'Ajou University': 2,
 'Alexandria University': 2,
 'Alexandru Ioan Cuza University': 1,
 'Aligarh Muslim University': 1,
 'American University': 1,
 'American University of Beirut': 2,
 'American University of Sharjah': 1,
 'Amirkabir University of Technology': 2,
 'Amrita University': 1,
 'Anadolu University': 1,
 'Andhra University': 1,
 'Aristotle University of Thessaloniki': 3,
 'Arizona State University': 3,
 'Asia University, Taiwan': 1,
 'Aston University': 2,
 'Athens University of Economics and Business': 1,
 'Auburn University': 3,
 'Auckland University of Technology': 1,
 'Austral University of Chile': 1,
 'Australian National University': 2,
 'Autonomous University of Barcelona': 3,
 'Autonomous University of Madrid': 3,
 'Babeş-Bolyai University': 1,
 'Ban

# 11. In the times ranking, compute the number of times each university appears

Determiniamo quante volte l'università è stata classificata nella tabella 'times_ranking'.

In [ ]:
t_r_appearance=times_ranking.groupby('uni_name_norm')['university_name'].count()

Successivamente, il dizionario 'uni_appearance', che contiene i nomi delle università e il rispettivo numero di classificazioni, lo utilizzo per per costruire il relativo dataframe, come si può vedere di seguito:

In [ ]:
t_r_appearance=pd.DataFrame(t_r_appearance.items(),columns=['university_name', 'appearance_num'])
t_r_appearance

university_name  appearance_num
0    AGH University of Science and Technology               1
1                          Aalborg University               5
2                            Aalto University               5
3                           Aarhus University               6
4                      Aberystwyth University               5
..                                        ...             ...
791               Yıldız Technical University               1
792                       Zhejiang University               6
793                  École Normale Supérieure              12
794                       École Polytechnique              12
795                         Örebro University               1

[796 rows x 2 columns]

# 12. Find the universities that appear at most twice in the times ranking

Ricaviamo dalla tabella 't_r_appearance' i nomi delle università che si ripetono almeno due volte nella colonna 'university_name' della tabella 'times_ranking'.

In [ ]:
t_r_appearance[t_r_appearance['appearance_num']<=2]

university_name  appearance_num
0    AGH University of Science and Technology               1
5                  Adam Mickiewicz University               1
6                    Aix-Marseille University               1
7                             Ajou University               1
9              Alexandru Ioan Cuza University               1
..                                        ...             ...
787                  Yokohama City University               1
788              Yokohama National University               1
790                        Yuan Ze University               2
791               Yıldız Technical University               1
795                         Örebro University               1

[394 rows x 2 columns]

#13. The universities that, in any year, have the same position in all three rankings (they must have the same position in a year)

In [ ]:
times_ranking['world_rank'] = times_ranking['world_rank'].str.replace('=','')

shanghai_ranking_reduced = shanghai_ranking[['uni_name_norm','year','world_rank']]

times_ranking_reduced = times_ranking[['uni_name_norm','year','world_rank']]

world_ranking_reduced = world_ranking[['uni_name_norm','year','world_rank']]

In [ ]:
intermediate_merge = pd.merge(shanghai_ranking_reduced,times_ranking_reduced,on=['uni_name_norm','year'])
same_position = pd.merge(intermediate_merge,world_ranking_reduced,on=['uni_name_norm','year'])
same_position.rename(columns={'world_rank_x':'shanghai_ranking',\
                                                   'world_rank_y':'times_ranking',\
                                                   'world_rank':'world_ranking',},inplace = True)

same_position[same_position['uni_name_norm'].isnull()==True]

Empty DataFrame
Columns: [uni_name_norm, year, shanghai_ranking, times_ranking, world_ranking]
Index: []

In [ ]:
same_position

uni_name_norm  year  ... times_ranking world_ranking
0                       Harvard University  2012  ...             2             1
1                      Stanford University  2012  ...             2             3
2    Massachusetts Institute of Technology  2012  ...             7             2
3       University of California, Berkeley  2012  ...            10            10
4                  University of Cambridge  2012  ...             6             4
..                                     ...   ...  ...           ...           ...
890                 University of Oklahoma  2015  ...       301-350           476
891                    University of Pavia  2015  ...       251-275           327
892                  University of Trieste  2015  ...       201-225           342
893        Vienna University of Technology  2015  ...       226-250           427
894                 Wake Forest University  2015  ...       201-225           159

[895 rows x 5 columns]

In [ ]:
same_position.dtypes

uni_name_norm       object
year                 int64
shanghai_ranking    object
times_ranking       object
world_ranking        int64
dtype: object

Definiamo la funzione possible match, con 3 condizioni:
 
*   il raking di shanghai e time contengono entrambi il range e confrontiamo i limiti superiori e inferiori tra i due range rispetto a world ranking.
*   solo per shanghai faccio il confronto con world ranking.
*   solo per times faccio il confronto con world ranking
*   Infine se i valori in shanghai times e world coincidono allora esiste il match







In [ ]:
def possible_match(db):
    
    #valori
  shanghai = db['shanghai_ranking']
  times = db['times_ranking']
  world = db['world_ranking']

  if ('-' in shanghai) and ('-' in times):
    pattern_1 = re.compile("(\d+)\-(\d+)")
    occurrence_1 = pattern_1.search(times)
    lower_1 = int(occurrence_1.group(1))
    upper_1 = int(occurrence_1.group(2))
                    
    pattern_2 = re.compile("(\d+)\-(\d+)")
    occurrence_2 = pattern_2.search(shanghai)
    lower_2 = int(occurrence_2.group(1))
    upper_2 =int(occurrence_2.group(2))
                
    if world >= lower_1 and world <= upper_1 and world >= lower_2 and world <= upper_2 :
      return 'possible'
    else:
      return 'not possible'

  elif  '-' in shanghai:
    pattern_3 = re.compile("(\d+)\-(\d+)")
    occurrence_3 = pattern_3.search(shanghai)
    lower_3 = int(occurrence_3.group(1))
    upper_3 = int(occurrence_3.group(2))
    if world >= lower_3 and world <= upper_3 and str(world) == times:
      return 'possible'
    else:
      return 'not possible'

  elif  '-' in times:
    pattern_4 = re.compile("(\d+)\-(\d+)")
    occurrence_4 = pattern_4.search(times)
    lower_4 = int(occurrence_4.group(1))
    upper_4 = int(occurrence_4.group(2))
    if world >= lower_4 and world <= upper_4 and str(world) == shanghai:
      return 'possible'
    else:
      return 'not possible'
  
  elif shanghai == times and shanghai == str(world) :
    return 'match'
  else:
    return 'no match'

In [ ]:
same_position['possible_match']=same_position.apply(possible_match, axis=1)

In [ ]:
final = same_position[(same_position['possible_match']=='possible') | (same_position['possible_match']== 'match')]	
final

uni_name_norm  year  ... world_ranking possible_match
80              Stanford University  2013  ...             2          match
280       University of Southampton  2014  ...           146       possible
318         University of Hong Kong  2014  ...           202       possible
408  Sungkyunkwan University (SKKU)  2014  ...           204       possible
468               Temple University  2014  ...           390       possible
486           University of Vermont  2014  ...           320       possible
709         University of Barcelona  2015  ...           241       possible
749            University of Vienna  2015  ...           265       possible
766                Laval University  2015  ...           229       possible
774            Newcastle University  2015  ...           212       possible
813          University of Waterloo  2015  ...           269       possible
855             University of Turku  2015  ...           383       possible
859     Washington State University  2015  ...           380       possible

[13 rows x 6 columns]